### ISSUE TITLE
Don't include urls with {} or similar patterns

### ISSUE THREAD

**Issue by maelle** [urlstechie/urlchecker-python#25]
Sorry I'm breaking my promise to no longer comment for a while!

I've run urlchecker via the action on a website, https://github.com/r-hub/docs/runs/576617965?check_suite_focus=true

I get false positives for e.g. `https://www.r-pkg.org/badges/version/` that's actually escaped in my content file (as code) https://github.com/r-hub/docs/blob/master/content/badges/_index.md#cran-versions I'm guessing it's because of regex?

To extract links (in R) I use the [commonmark package that uses cmark](https://cran.r-project.org/web/packages/commonmark/index.html). [Post about URL cleaning](https://ropensci.org/technotes/2019/12/19/urls-tidying/), [Post about commonmark](https://ropensci.org/technotes/2018/09/05/commonmark/#urls-parsing). There seems to be a [Python commonmark library](https://pypi.org/project/commonmark/). That said a limitation of this approach is that it wouldn't recognize URLs in text (that e.g. Hugo would transform to links).

**Comment by SuperKogito** [comment]
The link you mentioned https://www.r-pkg.org/badges/version/ is broken and your urlchecker logs show it. So it seems to me that the output delivered by urlchecker is correct.

![Screenshot_2020-04-10_14-12-01](https://user-images.githubusercontent.com/15731839/78989930-49cee380-7b35-11ea-9fa6-b2a51ddbab6c.png)

Tbh, regex is the standard/ the go to library in Python and I have never heard of commonmark before, so without a clear disadvantage in regex, I don't think we should replace it. I will keep this open, to let @vsoch weigh on this.

**Comment by maelle** [comment]
It is broken but it doesn't appear on my website: https://docs.r-hub.io/#badges

What I have there is `https://www.r-pkg.org/badges/version/{package}`

The false positive is not that bad, of course.

**Comment by SuperKogito** [comment]
I think I get your point now. You are saying that the link present in the file is `https://www.r-pkg.org/badges/version/{package}` but the urlchecker extracts this: `https://www.r-pkg.org/badges/version/`. If so, then yes, this is related to regex, the recognized structure does cover this link correctly because of `{` and `}`, are not recognized as part of links and as far as I know that should be the case.
I think in this case, it is better to whitelist the url or the pattern, unless urls with characters like `{` and `}` are frequent, then we can work on a fix for that.

**Comment by maelle** [comment]
More precisely, I am saying this is not an URL I really want to check because it is a template (the URL with curly braces doesn't exist either). In my website it appears as code. It could also happen when documenting something e.g. if in my docs I write `check('https://foo.com')` where foo.com is really just an example. 

I also realize that commonmark only applies to Markdown files whereas this module is more general (I wasn't thinking about this since I use it for Markdown files, sorry), so whitelisting makes sense.

**Comment by vsoch** [comment]
@maelle definitely no need to promise not to comment, your suggestions / bug reports help to make the library better, and I enjoy doing these fixes! I don't think I would have chosen being an RSE if I didn't :)

Let's walk through your use case (the repo https://github.com/r-hub/docs/runs/576617965?check_suite_focus=true) and talk about some of the issues you bring up. First I'll clone the repository:

```bash
git clone https://github.com/r-hub/docs
```

I'm going to do the very vanilla-est of runs - just a global check without any custom parameters.

```bash
urlchecker check .
```
I've easily reproduced your issue - it looks like there are urls in _index.md that are erroring:

![image](https://user-images.githubusercontent.com/814322/79013459-028d2680-7b26-11ea-927b-d09e6d9c38c3.png)

And you are correct that urlchecker doesn't have an understanding of commented out - it's just steamrolling through files, finding something that smells like a URL (based on a [regular expression](https://github.com/urlstechie/urlchecker-python/blob/master/urlchecker/core/urlmarker.py#L14) and continuing on happily. And there is something weird going on with those failed URLs *not* failing the job - that is a bug! We had some issues with a previous release and I bet that was introduced then. So I've opened an issue and I'll address this promptly. https://github.com/urlstechie/urlchecker-python/issues/27

But back to discussing these particular patterns - let's give a shot at first adding the `--files` argument so we can just run on that one file, here is what that looks like:

![image](https://user-images.githubusercontent.com/814322/79015486-120e6e80-7b2a-11ea-8f7e-234be59e8f36.png)

okay great! So now to debug why it's passing (when it should not). I found this bug too - it was an artifact of the previous release issue I mentioned. Not an issue, we will need to fix this asap. Okay, so you can see those two updates here (the fixed ending, and using --files)

https://asciinema.org/a/YaixINR7ddBYpRLVQEti6HCpx?speed=2

Now let's look into why the urls with {} are being parsed.

```python
from urlchecker.core.check import *                                                                                                    
file_path = "content/badges/_index.md"                                                                                                 

import os                                                                                                                              
os.path.exists(file_path)                                                                                                              
# True

urls = fileproc.collect_links_from_file(file_path)                                                                                     

['https://www.r-pkg.org/badges/version-ago/Rcpp',
 'https://www.r-pkg.org/badges/ago/',
 'https://www.r-pkg.org/badges/version/foobar',
 'https://cranlogs.r-pkg.org/badges/grand-total/Rcpp',
 'https://www.r-pkg.org/badges/version/',
 'https://www.r-pkg.org/badges/ago/Rcpp',
 'https://cranlogs.r-pkg.org/badges/last-month/Rcpp',
 'https://github.com/ropenscilabs/cchecksapi#badges',
 'https://cranlogs.r-pkg.org/badges/',
 'https://cranlogs.r-pkg.org/badges/last-week/Rcpp',
 'https://www.r-pkg.org/badges/version/Rcpp',
 'https://www.r-pkg.org/badges/version-last-release/Rcpp',
 'https://www.r-pkg.org/badges/last-release/',
 'https://github.com/r-hub/cranlogs.app#badges',
 'https://www.r-pkg.org/badges/version-last-release/',
 'https://cranlogs.r-pkg.org/badges/last-day/Rcpp',
 'https://www.r-pkg.org/badges/last-release/Rcpp',
 'https://www.r-pkg.org/badges/version-ago/']
```

Looking at your file, it does look like the regular expression is matching (partial) strings that don't include this last snippet, { }.

```
* `https://www.r-pkg.org/badges/version-ago/{package}` ![](https://www.r-pkg.org/badges/version-ago/Rcpp)
* `https://www.r-pkg.org/badges/version-last-release/{package}` ![](https://www.r-pkg.org/badges/version-last-release/Rcpp)
* `https://www.r-pkg.org/badges/ago/{package}` ![](https://www.r-pkg.org/badges/ago/Rcpp)
* `https://www.r-pkg.org/badges/last-release/{package}` ![](https://www.r-pkg.org/badges/last-release/Rcpp)
```
I'll need to think about how to address this - technically the urls (up to that point) are valid and matching, so the regular expression isn't accounting for that. I'm definitely familiar with commonmark, but I'm also conservative in adding new dependencies to any package, so I'll see if there is some way I can handle it natively first.

Lots of changes coming soon - stay tuned! And I haven't even looked at your other issues yet!

**Comment by vsoch** [comment]
@maelle I think what we will need to do is triage the other issues, and (for the time being) whitelist this file, and address the more difficult one in another PR.

**Comment by vsoch** [comment]
@SuperKogito - I think a good way to go about fixing this would be to have two regular expression passes:

  - the first gets a list that includes even ending urls that contain patterns like `{version}`
  - then we filter out those urls
  - then we run the current regexp, which should work the same (but not handle those urls).

Since you developed the original regex, do you want to take a shot at this?

**Comment by maelle** [comment]
Why not having users pass `https://www.r-pkg.org/badges/version/{package}` as a white-listed URL, then filter white-listed URLs from files, then apply your current regex? This way your regex doesn't need to account for weird URLs like mine? (I might be missing something!)

Thanks for the super thorough walk-through my repo in any case!

**Comment by vsoch** [comment]
The issue is that the regex is done first on all of the raw html or code content, so the ending variable isn’t included even in the first set to pass through whitelisting. The idea I was thinking of is to have two regex - the first is less stringent and would let that url through, then it could be whitelisted, and then the more stringent regex passed through the final set.

**Comment by maelle** [comment]
It makes sense! What I was imagining was using the whitelist as a regex filter on the raw content (a gsub or so), less elegant.

**Comment by vsoch** [comment]
okay @SuperKogito you should be unblocked to work on this now - the refactor of check (and organization of test file names, and writing a few new tests) was a large change, but I think we're ready if you want to work on this! If you don't have bandwidth let me know, I can give it a shot, but since you wrote the original regexp I will first leave to your expertise!

**Comment by SuperKogito** [comment]
This is going to be tricky to fix tbh. The original regex was done by someone much more experienced than me. I simply used his gist. Luckily I found some of his docs that might help, so I should be able to fix this. My only concern -in this case- is that hopefully there are no code languages "supported by our lib" that will have the following format `{url}`, otherwise this might backfire.  @vsoch  I will try to fix this today, if I cannot make it, I will let you know ;)

**Comment by vsoch** [comment]
There is no rush @SuperKogito ! The important thing is to do good, careful work, whatever time you need to do that.

**Comment by vsoch** [comment]
See https://github.com/urlstechie/urlchecker-python/pull/34#issuecomment-613721814 - we should have a solution! @maelle if you have time to take a look / review or test before we do a final merge, it would be great!

**Comment by SuperKogito** [comment]
This was fixed in #34.
